### PR TITLE
bin/util: make column families's options work

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Rocksdb has many different ways to open a db, let's have a look at the following two ways, 1) open db with the specified "name"; 2) open db with column families. The functions is declared as follows:
```
static Status Open(const Options& options, const std::string& name, DB** dbptr); 
static Status Open(const DBOptions& db_options, const std::string& name,
                                const std::vector<ColumnFamilyDescriptor>& column_families,
                                std::vector<ColumnFamilyHandle*>* handles, DB** dbptr);
```
Let's focus on the type of first parameter, they are different. Type Options is a wrapper, it is drived from DBOptions and ColumnFamilyOptions. DBOptions is used to control the behavor of db, like log directory, wal directory, max wal size, max background compaction threads and so on. ColumnFamilyOptions is used to control the behavor of column family, like write buffer size, compression type, condition of compaciton tigger, block cache size and so on.

Let‘s attention to rust-rocksdb library, it's DBOptions is Corresponds to Rocksdb's Options, and no type Corresponds to rocksdb's DBOptions and ColumnFamilyOptions. When we use rust-rocksdb's DBOptions, it can be passed to rocksdb as rocksdb's Options, DBOptions or ColumnFamilyOptions.

Take a look at how tikv open rocksdb, 1) try open_cf with cf_options not we provide, 2) if 1 successed, return, otherwise create db and all cfs with opts that we provide. When we start tikv with all cf are existed, we will drop options we have configed.

### CODE REVIEW

**Comment by ngaut** [comment]
LGTM

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -405,11 +440,14 @@ fn build_raftkv(matches: &Matches,
     let trans = ServerTransport::new(ch);
     let path = Path::new(&cfg.storage.path).to_path_buf();
     let opts = get_rocksdb_option(matches, config);
+    let cfs_opts = vec![get_rocksdb_option(matches, config), get_rocksdb_lock_cf_option()];
```
Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: use opts.clone() instead of call get_rocksdb_option again.

File: Cargo.lock [hunk 2]
```diff
@@ -192,7 +192,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 [[package]]
 name = "librocksdb_sys"
 version = "0.1.0"
-source = "git+https://github.com/ngaut/rust-rocksdb.git#fd83e59bda39f7ba5cddf94270e16343c44c6a8d"
+source = "git+https://github.com/ngaut/rust-rocksdb.git#561f1bf8bb9184727fa83d9ac62514b5f83a4a88"
```
Comment (BusyJay) [inline Cargo.lock]: You should update rocksdb url in `travis-build/Makefile` too.

File: src/util/rocksdb.rs [hunk 3]
```diff
@@ -52,11 +67,11 @@ pub fn new_engine_opt(mut opts: Options, path: &str, cfs: &[&str]) -> Result<DB,
         Ok(db) => db,
         Err(e) => return Err(e),
     };
-    for &cf in cfs {
+    for (&cf, &cf_opts) in cfs.iter().zip(cfs_ref_opts.iter()) {
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: just `cfs_ref_opts` is enough.

File: src/util/rocksdb.rs [hunk 4]
```diff
@@ -22,27 +22,42 @@ pub fn get_cf_handle<'a>(db: &'a DB, cf: &str) -> Result<&'a DBCFHandle, String>
 pub fn open(path: &str, cfs: &[&str]) -> Result<DB, String> {
     let mut opts = Options::new();
     opts.create_if_missing(false);
-    open_opt(&opts, path, cfs)
+    let mut cfs_opts = vec![];
```
Comment (siddontang) [inline src/util/rocksdb.rs]: can we use `vec![Options::new();cfs.len()]`?

Comment (zhangjinpeng1987) [inline src/util/rocksdb.rs]: Options can not be cloned

**Comment by disksing** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**